### PR TITLE
Log SMT formula to file for new incremental backend.

### PR DESCRIPTION
**Pull request by esteffin**
Add the possibility to log the SMT formula generated by the new SMT2 incremental backend to a file.

To use the feature add `--outfile <filename>` to command-line arguments.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-checker/solver_factory.cpp [hunk 1]
```diff
@@ -329,12 +332,30 @@ solver_factoryt::get_string_refinement()
...
+    out = util_make_unique<std::ofstream>(widen(outfile));
+#else
+    out = util_make_unique<std::ofstream>(outfile);
+#endif
+
+    if(!*out)
+    {
+      throw invalid_command_line_argument_exceptiont(
+        "failed to open file: " + outfile, "--outfile");
+    }
```
Comment (thomasspriggs) [inline src/goto-checker/solver_factory.cpp]: :pick: This section of code appears to have been duplicated from the `get_smt2`, can you separate it into a function. I guess the function should take the (potentially empty) `outfile` string and either throw or return an `std::unique_ptr<std::ofstream>`.

File: regression/cbmc-outfile/chain.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env python3
```
Comment (tautschnig) [inline regression/cbmc-outfile/chain.py]: Do we really need that script? Couldn't it also be done by using `--outfile /dev/stdout`?

Comment (esteffin) [reply regression/cbmc-outfile/chain.py]: Yes, unfortunately we need it as:

1. We test on windows where `/dev/stdout` may not work
2. By adding `--outfile /dev/stdout` we will know that the SMT formula is printed on the screen, not that it is added to the out file, so we cannot distinguish file from normal output.
3. The way the test output is checked is not in a certain order while the python script enforce a certain ordering, so if we want to have 2 `(check-sat)` calls the old system will be satisfied with one `(check-sat)` in the output, the new one instead will require 2 in the specified order

Comment (tautschnig) [reply regression/cbmc-outfile/chain.py]: 1. I believe you can use `-` instead of `/dev/stdout`, which will then work across platforms.
2. Granted, but I'm not sure this one aspect is a sufficient reason to warrant yet another piece of code that needs to be maintained.
3. You can use `activate-multi-line-match` to handle this case.

I don't intend to block on that, but every line of code that is written has a (future) cost. So I ask for careful consideration whether this really is needed.

Comment (esteffin) [reply regression/cbmc-outfile/chain.py]: Unfortunately in the new SMT backend we do not support the old `--outfile -` behaviour as, because of the interactive behaviour, the output will be interleaved with other messages (defying the sense of `--outfile`).
It is still possible to get the SMT formula on the `stdout` by using `--verbosity 10` and then filtering the lines sent to the solver.

Comment (tautschnig) [reply regression/cbmc-outfile/chain.py]: Hmm, good point. So I guess you have to keep the Python script, and I'm just hoping it won't break...

File: src/solvers/smt2_incremental/smt_solver_process.cpp [hunk 3]
```diff
@@ -28,6 +30,15 @@ void smt_piped_solver_processt::send(const smt_commandt &smt_command)
   const std::string command_string = smt_to_smt2_string(smt_command);
   log.debug() << "Sending command to SMT2 solver - " << command_string
               << messaget::eom;
+
+  if(out_stream != nullptr)
+  {
+    // Using std::endl to flush the stream as it is a debugging functionality,
+    // and we can guarantee a consistent output in case of hanging after
+    // (check-sat)
+    *out_stream << command_string << std::endl;
```
Comment (tautschnig) [inline src/solvers/smt2_incremental/smt_solver_process.cpp]: Do you perhaps just want to use `std::flush` instead? Most of the comment would remain useful to keep, though.

File: src/solvers/smt2_incremental/smt_solver_process.h [hunk 4]
```diff
@@ -35,9 +37,12 @@ class smt_piped_solver_processt : public smt_base_solver_processt
   ///   The command and arguments for invoking the smt2 solver.
   /// \param message_handler:
   ///   The messaging system to be used for logging purposes.
+  /// \param out_stream:
+  /// Pointer to the stream to print the SMT formula. `nullptr` if no output.
```
Comment (tautschnig) [inline src/solvers/smt2_incremental/smt_solver_process.h]: Indent

File: src/solvers/smt2_incremental/smt_solver_process.h [hunk 5]
```diff
@@ -35,9 +37,12 @@ class smt_piped_solver_processt : public smt_base_solver_processt
...
   smt_piped_solver_processt(
     std::string command_line,
-    message_handlert &message_handler);
+    message_handlert &message_handler,
+    std::unique_ptr<std::ostream> out_stream);
```
Comment (tautschnig) [inline src/solvers/smt2_incremental/smt_solver_process.h]: Is passing the `std::unique_ptr` by value rather than as a reference the right thing to do?

File: src/cbmc/cbmc_parse_options.cpp [hunk 6]
```diff
@@ -225,10 +225,12 @@ void cbmc_parse_optionst::get_command_line_options(optionst &options)
...
 
-  if(cmdline.isset("stop-on-fail") ||
-     cmdline.isset("dimacs") ||
-     cmdline.isset("outfile"))
+  if(
+    cmdline.isset("stop-on-fail") || cmdline.isset("dimacs") ||
+    (cmdline.isset("outfile") && !cmdline.isset("incremental-smt2-solver")))
+  {
     options.set_option("stop-on-fail", true);
+  }
```
Comment (tautschnig) [inline src/cbmc/cbmc_parse_options.cpp]: I'm either failing to parse the commit message or don't understand why the incremental SMT2 solver now interacts with "outfile" in this way. Is it the case that "outfile" has a very different semantics with the incremental SMT2 solver?! Does not seem like a good idea.

File: src/solvers/smt2_incremental/smt_solver_process.h (hunk 5 above)
Comment (esteffin) [reply src/solvers/smt2_incremental/smt_solver_process.h]: Yes as we want to extend the lifetime of the passed stream to the one of the `smt_solver_proces` object that is longer than the one of the `solver_factory` where the pointer is created.

File: src/solvers/smt2_incremental/smt_solver_process.cpp (hunk 3 above)
Comment (esteffin) [reply src/solvers/smt2_incremental/smt_solver_process.cpp]: As we want a new line and a `flush` I prefer `std::endl` instead of adding a new line + adding an explicit `std::flush`.

File: src/solvers/smt2_incremental/smt_solver_process.h (hunk 4 above)
Comment (esteffin) [reply src/solvers/smt2_incremental/smt_solver_process.h]: Fixed

File: src/cbmc/cbmc_parse_options.cpp (hunk 6 above)
Comment (thomasspriggs) [reply src/cbmc/cbmc_parse_options.cpp]: If we stop on fail, then we stop processing after the first round of solving. If we want to dump the SMT2 formula to file for all rounds of solving for ease of debugging (human reading) of the written file then we need to not stop on failure. Otherwise only the first round of solving is written to file, because we stopped on failure.

My somewhat related question would be why do we automatically stop on failure when `--outfile` is used with any of the decision procedures?

**Review by thomasspriggs — COMMENTED** [review]
Partial review only.

File: src/solvers/smt2_incremental/smt_solver_process.cpp (hunk 3 above)
Comment (NlightNFotis) [inline src/solvers/smt2_incremental/smt_solver_process.cpp]: Does it make sense to print a message that this action has been performed? E.g.

```
std::cout << "Outputting SMTLib to $file"
```

Probably somewhere else (constructor maybe?) so that the user can see what's going on (now I can use that, but there's no indication that this worked at all, which I found a bit surprising).

File: regression/cbmc-outfile/chain.py (hunk 2 above)
Comment (thomasspriggs) [reply regression/cbmc-outfile/chain.py]: > 1. I believe you can use `-` instead of `/dev/stdout`, which will then work across platforms.

The `-` option is special cased in the solver factory. So tests which rely on this aren't fully testing the code path for writing to file.

I think this python script has value beyond the new incremental smt backend, because it could also be used to test the dimacs output for example.

File: src/solvers/smt2_incremental/smt_solver_process.cpp (hunk 3 above)
Comment (tautschnig) [reply src/solvers/smt2_incremental/smt_solver_process.cpp]: Ok, but then you might want the comment to say "Using std::endl instead of \n"

File: src/cbmc/cbmc_parse_options.cpp (hunk 6 above)
Comment (tautschnig) [reply src/cbmc/cbmc_parse_options.cpp]: To your "somewhat related question:" no actual solving takes place with other back-ends with `--outfile`, and also: what would be the semantics of having only a single output file but multiple formulae having to be written?

So if the incremental SMT2 back-end has substantially different behaviour (solving actually takes place, multiple formulae are written) then we really shouldn't be using the same command-line option name. This confusing user-experience is a blocker.

Comment (peterschrammel) [reply src/cbmc/cbmc_parse_options.cpp]: >Otherwise only the first round of solving is written to file, because we stopped on failure.

But if you specified --stop-on-fail then this is exactly what you asked for. So, I don't see the point of having special behaviour here, @thomasspriggs. If you want all rounds to be written to the file then remove the --stop-on-fail flag.

Comment (thomasspriggs) [reply src/cbmc/cbmc_parse_options.cpp]: But this is for when we aren't explicitly specifying `--stop-on-fail`. The argument parsing is currently adding "stop-on-fail" implicitly when `--outfile` is specified. Do we need to add an explicit `--continue-on-fail` to override the implicit option?

Comment (peterschrammel) [reply src/cbmc/cbmc_parse_options.cpp]: I see. Then I'd call the current behaviour somewhat confusing... not sure how to fix that without having different behaviour for incremental vs non-incremental, or having multiple options. (Definitely **not** a --continue-on-fail option)

**Comment by thomasspriggs** [comment]
Re-opened as it was closed in error.

**Comment by esteffin** [comment]
Due to review comments, some extra work has been done so that now the `--outfile` works as with other backends (including `--outfile -` that will log on `stdout`.
Also when `--outfile` is added no SMT solver will be run.

There will be a subsequent PR that will add a new argument `--dump-smt-formula` that will log the SMT formula to a file while the SMT solver is also run (no `stop-on-fail` behaviour).

File: src/goto-checker/solver_factory.cpp [hunk 7]
```diff
@@ -148,7 +148,9 @@ std::unique_ptr<solver_factoryt::solvert> solver_factoryt::get_solver()
   const auto incremental_smt2_solver =
     options.get_option("incremental-smt2-solver");
   if(!incremental_smt2_solver.empty())
+  {
     return get_incremental_smt2(incremental_smt2_solver);
+  }
```
Comment (thomasspriggs) [inline src/goto-checker/solver_factory.cpp]: :-1: This formatting change is not related to the functional changes being made in this commit. I'd prefer that this was removed from this commit to avoid formatting related churn in the commit history.

File: src/solvers/smt2_incremental/smt_solver_process.cpp [hunk 8]
```diff
@@ -28,6 +28,7 @@ void smt_piped_solver_processt::send(const smt_commandt &smt_command)
   const std::string command_string = smt_to_smt2_string(smt_command);
   log.debug() << "Sending command to SMT2 solver - " << command_string
               << messaget::eom;
+
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_solver_process.cpp]: :-1: Formatting churn.

File: src/solvers/smt2_incremental/smt_solver_process.cpp [hunk 9]
```diff
@@ -68,3 +69,35 @@ smt_responset smt_piped_solver_processt::receive_response(
...
+}
+
+smt_responset smt_incremental_dry_run_solvert::receive_response(
+  const std::unordered_map<irep_idt, smt_identifier_termt> &identifier_table)
+{
+  // Using `smt_unsat_responset` as response because the decision-procedure will
+  // terminate anyway (stop-on-fail), it is not reported to the user as for
+  // `unknown`, and does not trigger a subsequent invocation to get the model
+  // (as a `smt_sat_responset` answer will trigger).
+  return *response_or_errort<smt_responset>{
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_solver_process.cpp]: :pick: I am not sure why you are constructing a `response_or_errort` here just to get the response out of it. Why not just construct and return the `smt_check_sat_responset`?

File: regression/cbmc-incr-smt2/nondeterministic-int-assert/stdout-match.desc [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+\(define-fun B0 \(\) Bool true\)
+\(define-fun B2 \(\) Bool \(not false\)\)
+\(assert B2\)
+\(assert \(= \(size_of_object \(_ bv1 8\)\) \(_ bv0 64\)\)\)
+\(assert \(= \(size_of_object \(_ bv0 8\)\) \(_ bv0 64\)\)\)
+\(check-sat\)
+^EXIT=0$
+^SIGNAL=0$
+--
+Test to check that the `--outfile -` argument is used correctly.
```
Comment (thomasspriggs) [inline regression/cbmc-incr-smt2/nondeterministic-int-assert/stdout-match.desc]: :pick: This file is missing a trailing newline character.

File: regression/cbmc-output-file/outfile/cvc5-match.desc [hunk 11]
```diff
@@ -0,0 +1,10 @@
+CORE
+test.c
+--incremental-smt2-solver 'cvc5 --lang=smtlib2.6 --incremental' --slice-formula --outfile %out-file-name%
+Passing problem to incremental SMT2 solving via SMT2 incremental dry-run
+Output file matches.
+^EXIT=0$
+^SIGNAL=0$
+--
+Test to check that the --outfile argument is used correctly and that the output file matches the
+rules defined in test-outfile-rules.txt
```
Comment (thomasspriggs) [inline regression/cbmc-output-file/outfile/cvc5-match.desc]: :pick: This file is missing a trailing newline character.

File: regression/cbmc-output-file/outfile/cvc5-no-match.desc [hunk 12]
```diff
@@ -0,0 +1,10 @@
+CORE
+test.c
+--incremental-smt2-solver 'cvc5 --lang=smtlib2.6 --incremental' --slice-formula --outfile %out-file-name%
+Passing problem to incremental SMT2 solving via SMT2 incremental dry-run
+Output file does not match.
+^EXIT=1$
+^SIGNAL=0$
+--
+Test to check that the --outfile argument is used correctly and that the output file matches the
+rules defined in test-outfile-rules.txt
```
Comment (thomasspriggs) [inline regression/cbmc-output-file/outfile/cvc5-no-match.desc]: :pick: This file is missing a trailing newline character.

File: regression/cbmc-output-file/outfile/z3-match.desc [hunk 13]
```diff
@@ -0,0 +1,10 @@
+CORE
+test.c
+--incremental-smt2-solver 'z3 --smt2 -in' --slice-formula --outfile %out-file-name%
+Passing problem to incremental SMT2 solving via SMT2 incremental dry-run
+Output file matches.
+^EXIT=0$
+^SIGNAL=0$
+--
+Test to check that the --outfile argument is used correctly and that the output file matches the
+rules defined in test-outfile-rules.txt
```
Comment (thomasspriggs) [inline regression/cbmc-output-file/outfile/z3-match.desc]: :pick: This file is missing a trailing newline character.

File: regression/cbmc-output-file/outfile/z3-no-match.desc [hunk 14]
```diff
@@ -0,0 +1,10 @@
+CORE
+test.c
+--incremental-smt2-solver 'z3 --smt2 -in' --slice-formula --outfile %out-file-name%
+Passing problem to incremental SMT2 solving via SMT2 incremental dry-run
+Output file does not match.
+^EXIT=1$
+^SIGNAL=0$
+--
+Test to check that the --outfile argument is used correctly and that the output file matches the
+rules defined in test-outfile-rules.txt
```
Comment (thomasspriggs) [inline regression/cbmc-output-file/outfile/z3-no-match.desc]: :pick: This file is missing a trailing newline character.

File: regression/cbmc-incr-smt2/nondeterministic-int-assert/stdout-match.desc [hunk 15]
```diff
@@ -0,0 +1,18 @@
+CORE
+test.c
+--outfile -
+Starting Bounded Model Checking
+\(set-option :produce-models true\)
+\(set-logic ALL\)
```
Comment (thomasspriggs) [inline regression/cbmc-incr-smt2/nondeterministic-int-assert/stdout-match.desc]: :bulb: I suggest surrounding the SMT formula regexes here with ^ and $. That way they match the full line, which differentiates the output from the verbose debug output. (I recognise that the debug output is already switched off on this test).

**Review by thomasspriggs — APPROVED** [review]
:+1:

File: src/goto-checker/solver_factory.cpp (hunk 7 above)
Comment (esteffin) [reply src/goto-checker/solver_factory.cpp]: It was part of older changes that are no more.
Removed.

File: src/solvers/smt2_incremental/smt_solver_process.cpp (hunk 8 above)
Comment (esteffin) [reply src/solvers/smt2_incremental/smt_solver_process.cpp]: It was part of older changes that will be added in the follow-up PR.
Removed.

File: src/solvers/smt2_incremental/smt_solver_process.cpp (hunk 9 above)
Comment (esteffin) [reply src/solvers/smt2_incremental/smt_solver_process.cpp]: Good point.
Changed.

File: regression/cbmc-incr-smt2/nondeterministic-int-assert/stdout-match.desc (hunk 10 above)
Comment (esteffin) [reply regression/cbmc-incr-smt2/nondeterministic-int-assert/stdout-match.desc]: Added.

File: regression/cbmc-output-file/outfile/cvc5-match.desc (hunk 11 above)
Comment (esteffin) [reply regression/cbmc-output-file/outfile/cvc5-match.desc]: Added.

File: regression/cbmc-output-file/outfile/cvc5-no-match.desc (hunk 12 above)
Comment (esteffin) [reply regression/cbmc-output-file/outfile/cvc5-no-match.desc]: Added.

File: regression/cbmc-output-file/outfile/z3-match.desc (hunk 13 above)
Comment (esteffin) [reply regression/cbmc-output-file/outfile/z3-match.desc]: Added.

File: regression/cbmc-output-file/outfile/z3-no-match.desc (hunk 14 above)
Comment (esteffin) [reply regression/cbmc-output-file/outfile/z3-no-match.desc]: Added.

File: regression/cbmc-incr-smt2/nondeterministic-int-assert/stdout-match.desc (hunk 15 above)
Comment (esteffin) [reply regression/cbmc-incr-smt2/nondeterministic-int-assert/stdout-match.desc]: Added.

**Review by thomasspriggs — APPROVED** [review]

**Comment by esteffin** [comment]
@tautschnig I think the new reworked commits should address your comments and concerns, especially as now the old `--outfile` behaviour has been restored.
Can you please double check this and if all good approve this PR?

**Review by peterschrammel — APPROVED** [review]
